c2rust-analyze/src/panic_detail.rs

```diff
@@ -0,0 +1,170 @@
+    backtrace: Option<Backtrace>,
+    sp: Span,
+}
+
+impl PanicDetail {
+    /// Create a new `PanicDetail` containing only a message, with no location or backtrace
+    /// information.
+    pub fn new(msg: String) -> PanicDetail {
+        PanicDetail {
+            msg,
+            loc: None,
+            relevant_loc: None,
+            backtrace: None,
+            sp: DUMMY_SP,
+        }
+    }
+
+    /// Returns `true` if this `PanicDetail` contains a backtrace.
+    pub fn has_backtrace(&self) -> bool {
+        self.backtrace.is_some()
+    }
+
+    /// Return a short (usually one-line) description of this panic.
+    pub fn to_string_short(&self) -> String {
+        let loc_str = self
+            .relevant_loc
+            .as_ref()
+            .or(self.loc.as_ref())
+            .map_or("[unknown]", |s| &*s);
+        format!("{}: {}", loc_str, self.msg.trim())
+    }
+
+    /// Return a full description of this panic, including a complete backtrace if available.
+    pub fn to_string_full(&self) -> String {
+        let mut s = String::new();
+        let loc_str = self.loc.as_ref().map_or("[unknown]", |s| &*s);
+        writeln!(s, "panic at {}: {}", loc_str, self.msg).unwrap();
+        if let Some(ref relevant_loc) = self.relevant_loc {
+            writeln!(s, "related location: {}", relevant_loc).unwrap();
+        }
+        if !self.sp.is_dummy() {
+            writeln!(s, "source location: {:?}", self.sp).unwrap();
+        }
+        if let Some(ref bt) = self.backtrace {
+            writeln!(s, "{:?}", bt).unwrap();
+        }
+        s
+    }
+}
+
+thread_local! {
+    static CURRENT_PANIC_DETAIL: Cell<Option<PanicDetail>> = Cell::new(None);
+}
+
+/// Panic hook for use with [`std::panic::set_hook`].  This builds a `PanicDetail` for each panic
+/// and stores it for later retrieval by [`take_current`].
+pub fn panic_hook(info: &PanicInfo) {
+    let bt = Backtrace::new();
+    let detail = PanicDetail {
+        msg: panic_to_string(info.payload()),
+        loc: info.location().map(|l| l.to_string()),
+        relevant_loc: guess_relevant_loc(&bt),
+        backtrace: Some(bt),
+        sp: CURRENT_SPAN.with(|cell| cell.get()),
+    };
+    let old = CURRENT_PANIC_DETAIL.with(|cell| cell.replace(Some(detail)));
+    if let Some(old) = old {
+        warn!("discarding old panic detail: {:?}", old);
+    }
+}
+
+/// Get the [`PanicDetail`] of the most recent panic.  This clears the internal storage, so if this
+/// is called twice in a row without an intervening panic, the second call always returns `None`.
+pub fn take_current() -> Option<PanicDetail> {
+    CURRENT_PANIC_DETAIL.with(|cell| cell.take())
+}
+
+/// Get the [`PanicDetail`] of the most recent panic; if it's not available, build a placeholder
+/// `PanicDetail` from the panic payload `e`.  This is useful in conjunction with
+/// [`std::panic::catch_unwind`].
+pub fn catch(e: &(dyn Any + Send + 'static)) -> PanicDetail {
```

Comment:
Could we just define a `panic_detail::catch_unwind` that wraps `std::panic::catch_unwind` with this?

Reply:
Good idea - done